### PR TITLE
Update of documentation

### PR DESCRIPTION
**Pull request by skovbakke**
An update to the documentation as well as documentation of newly implemented features and design changes.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,81 +1,86 @@
 # mukube
 
-Use [kas](https://kas.readthedocs.io/en/latest/userguide.html#usage) to build a yocto project.
+The aim of the project is to build a minimal linux distribution capable of runninb kubernetes and with support for needed security features. The project is based on [the yocto project](https://www.yoctoproject.org/) and [OpenEmbedded](http://www.openembedded.org/wiki/Main_Page), which uses [bitbake](https://www.yoctoproject.org/docs/latest/bitbake-user-manual/bitbake-user-manual.html) as the task execution engine. On top of that we use [kas](https://kas.readthedocs.io/en/latest/userguide.html#usage) to manage the project dependencies and it is also able to run the build.
```
Comment (klausenbusk) [inline README.md:3]:
```suggestion
The aim of the project is to build a minimal Linux distribution capable of running Kubernetes and with support for needed security features. The project is based on [the Yocto project](https://www.yoctoproject.org/) and [OpenEmbedded](http://www.openembedded.org/wiki/Main_Page), which uses [BitBake](https://www.yoctoproject.org/docs/latest/bitbake-user-manual/bitbake-user-manual.html) as the task execution engine. On top of that we use [kas](https://kas.readthedocs.io/en/latest/userguide.html#usage) to manage the project dependencies and build the project.
```

File: README.md [hunk 2]
```diff
@@ -1,81 +1,86 @@
...
 
+### Quick build
 * Create a folder for the build, e.g. `~/yocto`
 * Clone the repository into this folder, e.g. `~/yocto/mukube`
-* Invoke the script `kas checkout --update mukube/kas.yaml` from the `~/yocto` folder to setup the build environment.
-* Build the project using kas by running `kas build --update mukube/kas.yaml`.
+* Install have [kas](https://kas.readthedocs.io/en/latest/userguide.html) available and create a `kas-local.yaml`
```
Comment (klausenbusk) [inline README.md:8]:
```suggestion
* Install [`kas`](https://kas.readthedocs.io/en/latest/userguide.html#dependencies-installation) and create a `kas-local.yaml` file
```

File: README.md [hunk 3]
```diff
@@ -1,81 +1,86 @@
...
-To use the nfs server it must be mounted on the machine you are working on. 
-1. Install the package `nsf-common` using `apt`
-2. add the following line to `/etc/fstab` in order to mount the nfs at boot `<ip_of_nfs-server>:/	/mnt	nfs	auto	0	0` 
-
-An example local kas file which sets the downloads directory and sstate cache mirror is
+## Local configurations in `kas-local.yaml`
+A local kas file is _required_ for the kas tool to run, since we include it in our `kas-base.yaml` file. A minimal local kas file, which in turn does not change is: 
+```yaml
+header:
+	version: 10
```
Comment (klausenbusk) [inline README.md:22]:
```suggestion
  version: 10
```

File: README.md [hunk 4]
```diff
@@ -1,81 +1,86 @@
...
 
   SSTATE_MIRROR: | 
     SSTATE_MIRRORS ?= "\
-    file://.* file://<path to nfs sstate-cache folder>/PATH \
+    file://.* file://<path to nfs sstate-cache directory>/PATH \
     "
 ``` 
 
-To use the local kas file you have to chain them by `kas <command> mukube/kas.yaml:mukube/kas-local.yaml`.
+To use the nfs server, the exposed directories must be mounted on the machine you are working on.
```
Comment (klausenbusk) [inline README.md:53]:
```suggestion
To use a NFS server, the exposed directories must be mounted on the machine you are working on. 
```

File: README.md [hunk 5]
```diff
@@ -1,81 +1,86 @@
...
     SSTATE_MIRRORS ?= "\
-    file://.* file://<path to nfs sstate-cache folder>/PATH \
+    file://.* file://<path to nfs sstate-cache directory>/PATH \
     "
 ``` 
 
-To use the local kas file you have to chain them by `kas <command> mukube/kas.yaml:mukube/kas-local.yaml`.
+To use the nfs server, the exposed directories must be mounted on the machine you are working on. 
+1. Install the package `nsf-common` using eg. `apt`
+2. add the following line to `/etc/fstab` in order to mount the nfs at boot `<ip_of_nfs-server>:/	/mnt	nfs	auto	0	0`
```
Comment (klausenbusk) [inline README.md:55]:
```suggestion
1. Install the package `nfs-common` using eg. `apt`
2. Add the following line to `/etc/fstab` in order to mount the NFS at boot `<ip_of_nfs-server>:/	/mnt	nfs	auto	0	0` 
```

File: README.md [hunk 6]
```diff
@@ -1,81 +1,86 @@
...
     "
 ``` 
 
-To use the local kas file you have to chain them by `kas <command> mukube/kas.yaml:mukube/kas-local.yaml`.
+To use the nfs server, the exposed directories must be mounted on the machine you are working on. 
+1. Install the package `nsf-common` using eg. `apt`
+2. add the following line to `/etc/fstab` in order to mount the nfs at boot `<ip_of_nfs-server>:/	/mnt	nfs	auto	0	0` 
 
-As kas needs the local file to be in the same directory as the original kas file we require this file to be named `kas-local.yaml`
+## Building and Testing using kas/bitbake
```
Comment (klausenbusk) [inline README.md:57]:
```suggestion
## Building and testing using kas/bitbake 
```

File: README.md [hunk 7]
```diff
@@ -1,81 +1,86 @@
...
+To run the image locally, start by building the `dev` image, i.e. use the `kas-dev.yaml` as the target for the `kas build` command. Then from the `kas shell` (again with the `kas-dev.yaml` as the target) the image kan be launched in qemu with  
 
 `runqemu wic ovmf.secboot.code mukube nographic`
 
 *Be sure your user has access to `/dev/kvm` for [KVM](https://en.wikipedia.org/wiki/Kernel-based_Virtual_Machine) acceleration, on Ubuntu this can be done by running: `sudo usermod -a -G kvm $USER`.*
 
-### Launching the swtpm
-The image builds swtpm and there is a simple script to launch it. To run swtpm, launch another shell `kas shell mukube/<kas-file>` after building the image you want to test. Then launch the swtpm by running:
+### Tests 
+We have a few tests we run with the image. They are implemented in python are located in [cases](meta-k8s-setup/lib/oeqa/runtime/cases). Our test image is configured for the tests to run automatically when the target is built. Thus to run the tests, simply build the test target, with either `kas build mukube/kas-testing.yaml` or from the kas shell run `bitbake mukube-test-image`.
```
Comment (klausenbusk) [inline README.md:72]:
```suggestion
We have a few tests we run with the image. They are implemented in Python and are located in [`meta-k8s-setup/lib/oeqa/runtime/cases`](meta-k8s-setup/lib/oeqa/runtime/cases). Our test image is configured for the tests to run automatically when the target is built. Thus to run the tests, simply build the test target, with either `kas build mukube/kas-testing.yaml` or from the kas shell run `bitbake mukube-test-image`.
```

File: docs/design.md [hunk 8]
```diff
@@ -0,0 +1,14 @@
+# Design
+
+The following describes the components used in the image. 
+
+
+## Sedure Boot
```
Comment (klausenbusk) [inline docs/design.md:6]:
```suggestion
## Secure Boot
```

File: README.md [hunk 9]
```diff
@@ -1,81 +1,86 @@
...
-`bitbake mukube-test-image -c testimage ` 
-
-To clean the test build run:
-
-`bitbake -c clean mukube-test-image ` 
-
-For a full command that cleans, builds and runs the tests:
-
-`bitbake -c clean mukube-test-image && bitbake mukube-test-image && bitbake mukube-test-image -c testimage ` 
+Notice that the swtpm exits when the virtual machine turns off, and has to be relaunched before the next use.
```
Comment (klausenbusk) [inline README.md:83]:
```suggestion
Notice that swtpm exits when the virtual machine turns off, and has to be relaunched before the next use.
```

**Review by klausenbusk — CHANGES_REQUESTED** [review]

**Comment by skovbakke** [comment]
Did an almost full rewrite of the README page and included documentation of components of the boot process.

File: docs/image.md [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+
+On first boot [`systemd-repart`](https://www.freedesktop.org/software/systemd/man/systemd-repart.html) creates a encrypted partition for the Secure Boot keys (see [`50-volatile.conf`](../meta-k8s-setup/recipes-core/base-files/base-files/50-volatile.conf) and [man:repart.d(5)](https://www.freedesktop.org/software/systemd/man/repart.d.html)).
+
+## Secure Boot
+
+The image boots with UEFI Secure Boot. In a virtual environment UEFI is supported through ovmf ([open virtual machine firmware](https://www.linux-kvm.org/page/OVMF)). 
+
+## Unified kernel image, systemd-boot bootloader
+
+In order to simplify the Secure Boot process and have everything signed, we package the kernel, initrd, and the kernel command line in a [unified kernel image](https://systemd.io/BOOT_LOADER_SPECIFICATION#type-2-efi-unified-kernel-images). Booting this is supported with the bootloader [systemd-boot](https://www.freedesktop.org/software/systemd/man/sd-boot.html).
```
Comment (klausenbusk) [inline docs/image.md:15]:
```suggestion
In order to simplify the Secure Boot process and have everything signed, we package the kernel, initrd, and the kernel command line in a [Unified Kernel Image](https://systemd.io/BOOT_LOADER_SPECIFICATION#type-2-efi-unified-kernel-images). Booting this is supported with the bootloader [systemd-boot](https://www.freedesktop.org/software/systemd/man/sd-boot.html).
```

File: docs/image.md [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+
+A [partitioned image](https://docs.yoctoproject.org/dev-manual/common-tasks.html#creating-partitioned-images-using-wic) is created using `wic` (OpenEmbedded image creator). The partitions are described in [our kickstart file (`.wks`)](../meta-k8s-setup/wic/mukube.wks.in). 
+
+## Encrypted partition
+
+On first boot [`systemd-repart`](https://www.freedesktop.org/software/systemd/man/systemd-repart.html) creates a encrypted partition for the Secure Boot keys (see [`50-volatile.conf`](../meta-k8s-setup/recipes-core/base-files/base-files/50-volatile.conf) and [man:repart.d(5)](https://www.freedesktop.org/software/systemd/man/repart.d.html)).
+
+## Secure Boot
+
+The image boots with UEFI Secure Boot. In a virtual environment UEFI is supported through ovmf ([open virtual machine firmware](https://www.linux-kvm.org/page/OVMF)).
```
Comment (klausenbusk) [inline docs/image.md:11]:
```suggestion
The image boots with UEFI Secure Boot. In a virtual environment UEFI is supported through OVMF ([Open Virtual Machine Firmware](https://www.linux-kvm.org/page/OVMF)). 
```

File: README.md [hunk 12]
```diff
@@ -1,81 +1,86 @@
...
+To run the image locally, start by building the `dev` image, i.e. use the `kas-dev.yaml` as the target for the `kas build` command. Then from the `kas shell` (again with the `kas-dev.yaml` as the target) the image kan be launched in qemu with  
 
 `runqemu wic ovmf.secboot.code mukube nographic`
 
 *Be sure your user has access to `/dev/kvm` for [KVM](https://en.wikipedia.org/wiki/Kernel-based_Virtual_Machine) acceleration, on Ubuntu this can be done by running: `sudo usermod -a -G kvm $USER`.*
 
-### Launching the swtpm
-The image builds swtpm and there is a simple script to launch it. To run swtpm, launch another shell `kas shell mukube/<kas-file>` after building the image you want to test. Then launch the swtpm by running:
+### Tests 
+We have a few tests we run with the image. They are implemented in Python and are located in [`meta-k8s-setup/lib/oeqa/runtime/cases`](meta-k8s-setup/lib/oeqa/runtime/cases). Our test image is configured for the tests to run automatically when the target is built. Thus to run the tests, simply build the test target, with either `kas build mukube/kas-testing.yaml` or from the kas shell run `bitbake mukube-test-image`.
```
Comment (jacobcrawford) [inline README.md:72]: Have we actually made it so that the tests run on build? I did not enable the feature when I wrote the tests.

**Review by jacobcrawford — CHANGES_REQUESTED** [review]
A few comments. looks good.

File: README.md (hunk 12 above)
Comment (skovbakke) [reply README.md:72]: Yes, It is enabled now. I did that a few PRs back.

File: README.md [hunk 13]
```diff
@@ -1,81 +1,86 @@
...
 * Create a folder for the build, e.g. `~/yocto`
 * Clone the repository into this folder, e.g. `~/yocto/mukube`
-* Invoke the script `kas checkout --update mukube/kas.yaml` from the `~/yocto` folder to setup the build environment.
-* Build the project using kas by running `kas build --update mukube/kas.yaml`.
+* Install [`kas`](https://kas.readthedocs.io/en/latest/userguide.html#dependencies-installation) and create a `kas-local.yaml` file
+* Build the project using kas by running `kas build --update mukube/kas-base.yaml`
 
 ## Documentation
 
+* [Design](docs/design.md)
```
[the remark this answers is not in the document]
Comment (skovbakke) [reply README.md:13]: You are correct. That is a mistake.

**Review by jacobcrawford — APPROVED** [review]
LGTM

**Review by klausenbusk — COMMENTED** [review]
LGTM 🎉

**Review by klausenbusk — APPROVED** [review]
LGTM 🎉

**Pull request merged**